### ISSUE TITLE
BindKey() shouldn't be fired when the map engine is paused by another loop

### ISSUE THREAD

**Issue by DaVince** [spheredev/neosphere#26]
In Sir Boingers, I bind the key F7 to run "LEVEL.win()", which basically wins the level. It's basically a cheat. :)

The win() method has its own little loop that pauses the map engine so it can do its own thing before ending the level. However, BindKey() still seems to fire even though the map engine is paused, so if you hold F7 it will successively fire the win() method, even during this loop. That means you have to press space for every instance it opened before the game moves onward and ends the level properly.

The loop does have UpdateMapEngine() and RenderMap() but commenting those out makes zero difference.

**Comment by fatcerberus** [comment]
I knew this would come back to bite me eventually. :sweat_smile:

I deliberately set up BindKey to always fire as long as FlipScreen is called so that you could use it outside the map engine.  Sphere only checks it during the main MapEngine() loop.

As much as I like it this way, I can imagine it will break a lot of games, so I'll change it back to compatible behavior.

**Comment by DaVince** [comment]
Actually, I like your thinking there. Think you could add a new function for your version of the behavior instead?

**Comment by fatcerberus** [comment]
Something like `BindGlobalKey()` perhaps?

**Comment by DaVince** [comment]
Ha, that's exactly what I was thinking.

I'm also thinking that `BindKey()` would really have to be `BindMapKey()` then, with `BindKey()` being a direct reference to `BindMapKey()`... That would clarify things for new users, but probably actually just be bloat.

**Comment by fatcerberus** [comment]
Yeah, I don't like aliasing APIs like that if I can avoid it.  It's weird; most of Sphere's function names are completely self-documenting, except for BindKey (and the equivalent joystick function, BindJoystickButton).  The name suggests a global binding, when in reality it's only honored by the map engine.

Incidentally, it's this self-documenting property of the Sphere API which is why I love Sphere so much (I must, or I wouldn't have cared enough to make minisphere! :smile:) and the reason I have so many misgivings about the current state of the Pegasus spec.  There's this obsession with compartmentalizing everything that I think will put off newcomers.

**Comment by fatcerberus** [comment]
See these commits:
- https://github.com/fatcerberus/minisphere/commit/b38d5ea43037ac52840f612ad8f495cf0a950883
- https://github.com/fatcerberus/minisphere/commit/2c4cc122e270e119bbc3f71bd7aba791a9c96f3b

So it turns out you discovered a legitimate bug in the process: even though the global BindKey behavior was intentional, the recursion you saw wasn't.  BindKey scripts, like almost all event scripts in Sphere, aren't supposed to be reentrant.  The non-reentrancy is honored in 1.0, but it seems I broke it when I refactored the script handling recently.  Oops!

By the way, unless you're concerned with Sphere 1.5 compatibility, feel free to pass functions directly to any APIs normally expecting a code string such as `BindKey` or `SetUpdateScript`.  This allows you to use closures, which might be useful.
